superduperdb/backends/base/artifacts.py

```diff
@@ -108,47 +106,53 @@ def _save_file(self, file_path: str, file_id: str) -> str:
         """Save file in artifact store and return file_id."""
         pass
 
+    #
     def save_artifact(self, r: t.Dict):
-        """
-        Save serialized object in the artifact store.
+        """Save serialized object in the artifact store.
 
         :param r: dictionary with mandatory fields
-                  {'bytes', 'datatype'}
-                  and optional fields
-                  {'file_id', 'uri'}
-        """
-        if (
-            r.get('artifact_type') != 'file'
-            and 'file_id' in r
-            and r.get('bytes') is None
-        ):
-            return
-        if r.get('artifact_type') == 'file':
-            assert 'file_id' in r, 'file_id is missing!'
-            file_id = self._save_file(r['uri'], r['file_id'])
-        else:
-            assert 'bytes' in r, 'serialized bytes are missing!'
-            assert 'datatype' in r, 'no datatype specified!'
-            datatype = self.serializers[r['datatype']]
-            uri = r.get('uri')
-            file_id = r.get('file_id')
-            if uri is not None:
-                file_id = _construct_file_id_from_uri(uri)
-            else:
-                file_id = r.get('sha1') or hashlib.sha1(r['bytes']).hexdigest()
-            if r.get('directory'):
-                file_id = os.path.join(datatype.directory, file_id)
+        """
+        blobs = r.get('_blobs', {})
+        files = r.get('_files', {})
+
+        for file_id, blob in blobs.items():
+            try:
+                self._save_bytes(blob, file_id=file_id)
+            except FileExistsError:
+                continue
+
+        for file_id, file_path in files.items():
             try:
-                self._save_bytes(r['bytes'], file_id=file_id)
+                self._save_file(file_path, file_id=file_id)
             except FileExistsError:
-                logging.warn(
-                    f'Artifact with file_id {file_id} already exists, skipping...'
-                )
+                continue
 
-            del r['bytes']
-        r['file_id'] = file_id
+        r['_blobs'] = list(blobs.keys())
+        r['_files'] = list(files.keys())
         return r
 
+    def delete_artifact(self, r: t.Dict):
```

Comment:
I'm not really sure we need this method. 
Since the artifacts and files are flat in `Document.encode`, we can 
just directly with `save_bytes`, `save_file`.

Reply:
Not a big deal however - can be refactored later.